File: src/test/java/com/uapp_llc/test/stub/repository/IdentifyingRepositoryStub.java
package com.uapp_llc.test.stub.repository;

import java.util.Objects;

import com.uapp_llc.test.stub.repository.identification.Identification;

abstract class IdentifyingRepositoryStub<ID, T> extends AbstractRepositoryStub<ID, T> {

  private final Identification<T> identification;

  IdentifyingRepositoryStub(Identification<T> identification) {
    this.identification = identification;
  }

  protected abstract ID getId(T entity);

  @Override
  public T save(ID id, T entity) {
    Objects.requireNonNull(entity, "Entity must not be null");

    if (id == null || !super.exists(id)) {
      identification.apply(entity);
      ID entityId = this.getId(entity);
      logger.debug("Applied id '{}' to entity: {}", entityId, entity);
      return super.save(entityId, entity);
    }

    return super.save(id, entity);
  }

  public T save(T entity) {
    ID id = this.getId(entity);
    return this.save(id, entity);
  }

  public void delete(T entity) {
    Objects.requireNonNull(entity, "Entity must not be null");
    ID id = this.getId(entity);
    super.delete(id, entity);
  }

}

File: src/test/kotlin/com/uapp_llc/test/stub/repository/IdentifyingRepositoryStub.kt
package com.uapp_llc.test.stub.repository

import com.uapp_llc.test.stub.repository.identification.Identification

abstract class IdentifyingRepositoryStub<ID, T>(

    private val identification: Identification<T>

) : AbstractRepositoryStub<ID, T>() {

  protected abstract fun getId(entity: T): ID?

  override fun save(id: ID?, entity: T): T {
    if (id == null || !super.exists(id)) {
      identification.apply(entity)
      val entityId: ID? = this.getId(entity)
      logger.debug("Applied id '{}' to entity: {}", entityId, entity)
      return super.save(entityId, entity)
    }

    return super.save(id, entity)
  }

  open fun save(entity: T): T {
    val id: ID? = this.getId(entity)
    return this.save(id, entity)
  }

  open fun delete(entity: T) {
    val id: ID? = this.getId(entity)
    super.delete(id, entity)
  }

}
